### PR TITLE
Invalidating cache  for user/{user_id} endpoint with modifications for the GET, PUT and DELETE method

### LINKED ISSUE #191: [Task] Expand user/{user_id} redis cache to invalidate duplicate/invalid cache
**Is your feature request related to a problem? Please describe.**
This task requires code to assist with invalidating the cache (replace/deleting).  In addition, we need to make sure that there is no duplicate user_id and that you cannot delete user_id  that is not there.

**Describe the solution you'd like**
I wrote a "validator" method for the invalidation process.  This method would check to see if  the redis is cachable or is clearable before replacing it or deleting it.

**Describe alternatives you've considered**
N/A

**Additional context**
This is an extension of issue #138. In addition, this was a (unintentional) collaboration of two individuals.  I also updated the REDIS_HOST and REDIS_PORT to be environmental variables. 

TESTING DOCUMENTATION (copied and pasted from PR #231)

Before you test, please make sure under 'REDIS_HOST' and 'REDIS_PORT', please specify which ports and hosts you would like to use

1.)If you have VSC code open and navigated to your function, go ahead and run/debug
2.) Open Postman
3.) Run GET
4.) Your expected output on VSC should be the logs and that the results were all cached
5.)Run PUT
6.) Your expected output on VSC should be the same as above, except you will get a message that the cache was updated/unable to update, and a user_id should automatically generate in that SQL table.
a.) Enter whatever you like for first name, last name, email address
b.) You could also make an incomplete entry (like missing first name, email) to see if you get an error message.
7.) Run DELETE
8.)You should get a message that your cache has been cleared or deletion was failed.

Special thanks to PR #223 to help me with working on these documentation.

### PR DESCRIPTION
**Pull request by chonger878**
Closes #191 

This PR:
-Includes modification for PUT and DELETE methods for the user/{user_id} endpoint
-Includes a canValidate() method that can ensure the user is not deleting/replacing data from a cache that is empty/has duplicate values
-I have turned the REDIS_HOST and REDIS_PORT into os.environ variables so that different ports/hosts can use this- no longer hardcoded values.

This was the work of two contributors.

***TESTING DOCUMENTATION***

Before you test, please make sure under 'REDIS_HOST' and 'REDIS_PORT', please specify which ports and hosts you would like to use

1.)If you have VSC code open and navigated to your function, go ahead and run/debug
2.) Open Postman
3.) Run GET
4.) Your expected output on VSC should be the logs and that the results were all cached
5.)Run PUT
6.)  Your expected output on VSC should be the same as above, except you will get a message that the cache  was updated/unable to update, and a user_id should automatically generate in that SQL table.
     a.) Enter whatever you like for first name, last name, email address
     b.) You could also make an incomplete entry (like missing first name, email) to see if you get an error message.
7.) Run DELETE
8.)You should get a message that your cache has been cleared or  deletion was failed.

Special thanks to PR #223 to help me with working on these documentations.

### CODE REVIEW

**Review by kieranmiguel — APPROVED** [review]

File: functions/HttpTriggerAPIUsersId/__init__.py [hunk 1]
```diff
@@ -5,7 +5,7 @@
 import json
 import redis
 
-
+ALL_USERS_KEY = b'users:all'
```
Comment (farhadbahrehmandhenry) [inline functions/HttpTriggerAPIUsersId/__init__.py:9]: `ALL_USERS_KEY = b'user'
`
also add 
`CACHE_TOGGLE = os.environ["CACHE_TOGGLE"]
`

File: functions/HttpTriggerAPIUsersId/__init__.py [hunk 2]
```diff
@@ -78,25 +78,27 @@ def get_db_connection():
...
 
 def get_user(conn, user_id, _redis):
-    try:
+    is_cachable = False
+    try: 
         cache = get_user_cache(_redis)
         user = json.loads(cache)
         user_user_id = user['userId']
-        is_cachable = cache is not None and int(user_user_id) == int(user_id)
+        is_cachable = canInvalidate(user, user_id)
```
Comment (farhadbahrehmandhenry) [inline functions/HttpTriggerAPIUsersId/__init__.py]: I think it's a good idea to clear the cache if it's not cachable. add this after this line
```
        if not is_cachable:
            clear_cache(_redis)
```

File: functions/HttpTriggerAPIUsersId/__init__.py [hunk 3]
```diff
@@ -78,25 +78,27 @@ def get_db_connection():
...
     except TypeError as e:
         logging.info(e.args[0])
 
     try:
-        if (cache is None) or not is_cachable:
+        if (cache is None) or is_cachable:
```
Comment (farhadbahrehmandhenry) [inline functions/HttpTriggerAPIUsersId/__init__.py]: this line should be 
`if (cache is None) or not is_cachable:`

Comment (farhadbahrehmandhenry) [reply functions/HttpTriggerAPIUsersId/__init__.py]: also, add this after this line to handle CACHE TOGGLE
```
            if (CACHE_TOGGLE == "On"):
                logging.info("Cache is empty, querying database...")
```

Comment (farhadbahrehmandhenry) [reply functions/HttpTriggerAPIUsersId/__init__.py]: also wrap cache_user(_redis, user) in around line 137 in an if statement like this =>
```
                if (CACHE_TOGGLE == "On"):
                    # Cache the results
                    cache_user(_redis, user)
```

Comment (farhadbahrehmandhenry) [reply functions/HttpTriggerAPIUsersId/__init__.py]: the if statement in around line 144 should be elif not if.

`        elif cache is not None:
`

File: functions/HttpTriggerAPIUsersId/__init__.py [hunk 4]
```diff
@@ -147,7 +149,7 @@ def get_user(conn, user_id, _redis):
 
 def get_user_cache(_redis):
```
Comment (farhadbahrehmandhenry) [inline functions/HttpTriggerAPIUsersId/__init__.py:162]: in order to handle CACHE_TOGGLE you should wrap the body like this in an if statement =>
```
    if (CACHE_TOGGLE == "On"):
        try:
            cache = _redis.get(ALL_USERS_KEY)
            return cache
        except TypeError as e:
            logging.critical('Failed to fetch user from cache: ' + e.args[1])
            return None
    else:
        return None
```

File: functions/HttpTriggerAPIUsersId/__init__.py [hunk 5]
```diff
@@ -176,6 +178,7 @@ def update_user(user_req_body, conn, user_id, _redis):
         assert 'lastName' in user_req_body, 'User request body did not contain field: "lastName"'
         assert 'email' in user_req_body, 'User request body did not contain field: "email"'
     except AssertionError as user_req_body_content_error:
+        logging.error(e.args[0])
```
Comment (farhadbahrehmandhenry) [inline functions/HttpTriggerAPIUsersId/__init__.py]: logging.error(user_req_body_content_error.args[0])

File: functions/HttpTriggerAPIUsersId/__init__.py [hunk 6]
```diff
@@ -323,3 +326,10 @@ def get_user_req_body(req):
         pass
 
     return user_req_body
+
+def canInvalidate(cache, user_id):
```
Comment (farhadbahrehmandhenry) [inline functions/HttpTriggerAPIUsersId/__init__.py]: you can just return this => 

`    return cache is not None and int(user_user_id) == int(user_id)
`

File: functions/HttpTriggerAPIUsersId/__init__.py (hunk 1 above)
Comment (chonger878) [reply functions/HttpTriggerAPIUsersId/__init__.py:9]: Cache_toggle not needed for this sprint.

File: functions/HttpTriggerAPIUsersId/__init__.py (hunk 2 above)
Comment (chonger878) [reply functions/HttpTriggerAPIUsersId/__init__.py]: Done.  Thanks for the input!

File: functions/HttpTriggerAPIUsersId/__init__.py (hunk 5 above)
Comment (chonger878) [reply functions/HttpTriggerAPIUsersId/__init__.py]: Done.  Thanks!

File: functions/HttpTriggerAPIUsersId/__init__.py (hunk 6 above)
Comment (chonger878) [reply functions/HttpTriggerAPIUsersId/__init__.py]: Done. Thanks for that!

File: functions/HttpTriggerAPIUsersId/__init__.py (hunk 3 above)
Comment (chonger878) [reply functions/HttpTriggerAPIUsersId/__init__.py]: I fixed that particular if statement. :)

**Review by farhadbahrehmandhenry — APPROVED** [review]

**Pull request merged**